(this["webpackJsonpphylo-lib-web"]=this["webpackJsonpphylo-lib-web"]||[]).push([[0],{141:function(t,a,e){t.exports=e(201)},146:function(t,a,e){},201:function(t,a,e){"use strict";e.r(a);var n=e(0),c=e.n(n),r=e(38),g=e.n(r),o=(e(146),e(99)),i=e(63),l=e(28),s=(e(64),e(208)),A=e(216),u=e(213),G=e(23),m=e(206),T=e(207),f=e(209),h=e(108),d=e(202),C=e(203),E=e(211);function p(t){var a=t.content,e=t.vocab,n=t.link,r=t.linkText;return c.a.createElement(s.a,{className:"hover-vocab",content:c.a.createElement("div",null,a,c.a.createElement("p",null,"See"," ",c.a.createElement("a",{href:n,target:"_blank",rel:"noopener noreferrer"},r||"Wikipedia")," ","for more details.")),placement:"top",trigger:"hover"},c.a.createElement("div",null,e))}var v="\n\nC\n|\no-SARS\n|\no-C\n  |\n  o-C\n  | |\n  | o-H6N1\n  | |\n  | o-C\n  |   |\n  |   o-H5N1\n  |   |\n  |   o-C\n  |     |\n  |     o-H3N2\n  |     |\n  |     o-C\n  |       |\n  |       o-H7N9\n  |       |\n  |       o-C\n  |         |\n  |         o-H1N1\n  |         |\n  |         o-H7N7\n  |\n  o-C\n    |\n    o-C\n    | |\n    | o-H2N3\n    | |\n    | o-H10N7\n    |\n    o-C\n      |\n      o-H3N1\n      |\n      o-H3N8\n\n",b="\nC\n|\no-C\n| |\n| o-FOLSOMIA PENICULA\n| |\n| o-C\n|   |\n|   o-ANOPHELES GABONENSIS\n|   |\n|   o-C\n|     |\n|     o-BRACHYTELES HYPOXANTHUS\n|     |\n|     o-C\n|       |\n|       o-PARIDES ANCHISES\n|       |\n|       o-PARIDES NEOPHILUS\n|\no-C\n  |\n  o-C\n  | |\n  | o-CONOGETHES PUNCTIFERALIS\n  | |\n  | o-GRAPHOLITA MOLESTA\n  |\n  o-C\n    |\n    o-PARIDES PRONEUS\n    |\n    o-PYTHIUM INSIDIOSUM\n",y="\nC\n|\no-PEPPER VEINAL MOTTLE VIRUS\n|\no-C\n  |\n  o-C\n  | |\n  | o-BEAK AND FEATHER DISEASE\n  | |\n  | o-C\n  |   |\n  |   o-AVIAN NEPHRITIS VIRUS\n  |   |\n  |   o-C\n  |     |\n  |     o-C\n  |     | |\n  |     | o-C\n  |     | | |\n  |     | | o-DENGUE VIRUS\n  |     | | |\n  |     | | o-C\n  |     | |   |\n  |     | |   o-BOVINE VIRAL DIARRHEA VIRUS\n  |     | |   |\n  |     | |   o-HUMAN RHINOVIRUS\n  |     | |\n  |     | o-C\n  |     |   |\n  |     |   o-NOROVIRUS\n  |     |   |\n  |     |   o-C\n  |     |     |\n  |     |     o-FELINE IMMUNODEFICIENCY VIRUS\n  |     |     |\n  |     |     o-REDSPOTTED GROUPER NERVOUS NECROSIS VIRUS\n  |     |\n  |     o-C\n  |       |\n  |       o-HEPATITIS E\n  |       |\n  |       o-C\n  |         |\n  |         o-HEPATITIS C\n  |         |\n  |         o-NORWALK VIRUS\n  |\n  o-C\n    |\n    o-EASTERN EQUINE ENCEPHALITIS VIRUS\n    |\n    o-C\n      |\n      o-JAPANESE ENCEPHALITIS VIRUS\n      |\n      o-C\n        |\n        o-COXSACKIEVIRUS\n        |\n        o-SINDBIS VIRUS STRAIN\n",w=m.a.Content;function _(){var t=Object(n.useState)(b),a=Object(G.a)(t,2),e=a[0],r=a[1],g=Object(n.useState)(!0),o=Object(G.a)(g,2),l=o[0],s=o[1];return console.log("/phylo-lib-web"),c.a.createElement("div",{className:"wrapper"},c.a.createElement(w,{justify:"center"},c.a.createElement("div",{className:"home"},c.a.createElement(T.a,{className:"intro",justify:"center"},c.a.createElement("div",null,c.a.createElement("h1",null,"Welcome to PhyloML"),c.a.createElement("h2",null,"A phylogenetic library written in OCaml."),c.a.createElement("h2",null,"Inferring Evolutionary History through Modern Genetic Similarity."))),c.a.createElement(T.a,{justify:"center",gutter:[16,40]},c.a.createElement(f.a,{lg:7,md:10,sm:12},c.a.createElement(i.b,{className:"offset",to:"/pairwise"},"Visualize Pairwise DNA Alignments")),c.a.createElement(f.a,{lg:7,md:10,sm:12},c.a.createElement(i.b,{className:"offset",to:"/visualize"},"Visualize Phylogenetic Trees")),c.a.createElement(f.a,{lg:7,md:10,sm:12},c.a.createElement(i.b,{className:"offset",to:"/generate"},"Generate Phylogenetic Trees"))),c.a.createElement(T.a,{className:"horizontally-centered home-down"},c.a.createElement("a",{href:"#explanation"},c.a.createElement(h.a,{shape:"circle",icon:c.a.createElement(E.a,null),size:"large",style:{backgroundColor:"#c3efd9",borderColor:"#000000",color:"#000000"}})))),c.a.createElement("div",{className:"explanation",id:"explanation"},c.a.createElement(T.a,{justify:"center"},c.a.createElement(d.a,null),c.a.createElement("h1",null," Explanations and Examples ")),c.a.createElement(T.a,null,c.a.createElement("h2",null," About Phylogenetic Trees: "),c.a.createElement("h3",null," ","Scientists often wish to infer the evolutionary history between different organisms. Historically, in order to determine the closeness of species such as birds or fish, the similarity in physical characteristics was analyzed. With the advent of modern computers and DNA analysis, the field has moved towards using DNA similarity. Two species with more similar DNA are assumed to be more closely related. By analyzing these similarities and differences in the DNA, we can generate a hypothetical evolutionary tree, called a"," ",c.a.createElement(p,{content:c.a.createElement("p",null,"A tree that displays the evolutionary relationships between various organisms."),vocab:"phylogenetic tree",link:"https://www.khanacademy.org/science/high-school-biology/hs-evolution/hs-phylogeny/a/phylogenetic-trees",linkText:"Khan Academy"}),", that estimates what the historical evolutionary tree would have looked like."),c.a.createElement("h2",null," Functionality: "),c.a.createElement("h3",null,c.a.createElement("b",null," Visualize Pairwise Alignment: "),"At the heart of constructing phylogenetic trees is sequence alignment, where gaps are inserted in the sequences to make them all of equal length, to get a similarity score. Only once the alignment scores of all DNA sequences with one another is known is it possible to determine genetic similarity and through that, evolutionary history.",c.a.createElement("br",null),"One such method is the pairwise alignment, where two DNA sequences are aligned. It is performed using the Needleman-Wunsch algorithm, a globally optimal dynamic programming algorithm that computes a loss for the most optimal alignment. Here"," ",c.a.createElement("span",{className:"highlight"},"*")," represents a DNA base pair match, ",c.a.createElement("span",{className:"highlight"},"_")," a gap, often used if a DNA sequence is shorter than the other, and"," ",c.a.createElement("span",{className:"highlight"},"|")," a mismatch."),c.a.createElement("br",null),c.a.createElement("h3",null,c.a.createElement("b",null," Visualize Phylogenetic Trees: "),"When a phylogenetic tree has already been constructed, whether through this program or other commercially available software, it is important that the user can visualize the final alignment. Here we built a tokenizer, lexer, and parser to read PhyloXML, an extension of the popular XML format designed to encode phylogenetic trees."),c.a.createElement("br",null),c.a.createElement("h3",null,c.a.createElement("b",null," Generate Phylogenetic Trees: "),"We construct phylogenetic genetic trees from DNA sequences encoded in a standard .FASTA format using two algorithms: Unweighted Pair Group Method with Arithmetic Mean (UPGMA) and Bayesian Inference, each with their respective tradeoffs regarding speed and accuracy. Each algorithm reads in the .FASTA files, constructs pairwise alignments, and then uses respective techniques to find an optimal rooted tree.",c.a.createElement("br",null),"The first step in"," ",c.a.createElement(p,{vocab:"UPGMA",link:"https://en.wikipedia.org/wiki/UPGMA",linkText:"the Wikipedia page"})," ","consists of creating an ",c.a.createElement("i",null,"n")," by ",c.a.createElement("i",null,"n")," distance matrix which records the"," ",c.a.createElement(p,{vocab:"genetic distance",link:"https://en.wikipedia.org/wiki/Genetic_distance",linkText:"the Wikipedia page on genetic distance measures"})," ","between each pair of sequences in the alignment. Then, we pick the 2 sequeneces with the least distance and combine them to form a 'clade' (a branching point in the tree). Then we do the same thing with the next two most similar sequences. We keep doing this until we are left with one clade - our final phylogenetic tree."," ",c.a.createElement("br",null),"The essence of"," ",c.a.createElement(p,{vocab:"Bayesian inference",link:"https://www.sciencemag.org/site/feature/data/1050262.pdf",linkText:"a paper on phylogenetic Bayesian inference"})," ","is using prior information about the genetic data to predict the likelihood that a certain tree is the correct phylogenetic tree, given our aligned sequence data. Since finding this likelihood for every possible tree structure takes an unfeasibly large amount of time, we use a technique called"," ",c.a.createElement(p,{vocab:"Markov Chain Monte Carlo sampling",link:"http://pages.stat.wisc.edu/~larget/phylogeny/larget-simon-MBE-1999.pdf",linkText:"a paper on phylogenetic MCMC methods"})," ","to construct a chain of trees and get a sample of the final tree distribution. By getting a sample of the tree distribution, we can find out the most likely phylogenetic tree given our genetic data. The advantage of this method is that it can be tweaked considerably to account for various genetic substituion models, based on the existing knowledge that we have about the genetic sequences. It is, however, slower than UPGMA.")),c.a.createElement("h2",null," Examples: "),c.a.createElement("h3",null,"Examples of constructed phylogenetic trees using DNA sequences from the National Center for Biotechnology Information (NCBI)."),c.a.createElement(T.a,{className:"centered-content"},c.a.createElement(C.default.Group,{onChange:function(t){switch(t.target.value){case"Influenza A Viruses":r(v);break;case"COXII":r(b);break;case"Virus Capsid":s(!0),console.log(y),r(y)}},defaultValue:"COXII"},c.a.createElement(C.default.Button,{value:"Virus Capsid"},"Virus Capsid Gene"),c.a.createElement(C.default.Button,{value:"COXII"},"Cytochrome C Oxidase Subunit II"),c.a.createElement(C.default.Button,{value:"Influenza A Viruses"},"Influenza A PB-2"))),l?c.a.createElement(T.a,{justify:"center"},c.a.createElement("div",{className:"ascii-phylo-container"},c.a.createElement("p",{className:"ascii-phylo"},e))):null)))}var N=e(105),I=e.n(N),k=e(125),S=e(210),x=e(205),O=e(204),R=e(87),D=e(212),j=e(214),P=e(215),X=e(20),M=e(52),H=e(21),B=e(25),F=e(14),L=e(7),U=e(32),z=e(13);U.b("Dna-PhyloLibWeb.Empty");function V(t){return">"===F.h(t,0,1)}function q(t,a){var e=F.i(t);return F.g((function(t){if(function(t){if(t>=72)return 84===t||95===t;if(45===t)return!0;if(t<65)return!1;switch(t-65|0){case 1:case 3:case 4:case 5:return!1;case 0:case 2:case 6:return!0}}(t)||function(t){var a=t-65|0;return a>13||a<0?!(a>25||a<15):9!==a}(t))return H.a(a,t)}),e)}var W=H.c;function J(t){var a;try{a=F.e(t,10)}catch(c){var e=z.a(c);if("Not_found"!==e.RE_EXN_ID)throw e;a=0}var n=F.h(t,0,a);return""===n||V(n)?F.h(t,a+1|0,(t.length-a|0)-1|0):t}function K(t){var a=H.d(128);return q(J(t),a),a}var Y=H.e;function Q(t){var a=function(t,a){for(;;){var e=a,n=t;if(!V(n))return e;var c,r=J(n);try{c=F.e(r,62)}catch(s){var g=z.a(s);if("Not_found"!==g.RE_EXN_ID)throw g;c=r.length-1|0}var o,i=F.h(r,0,c);try{o=F.h(r,c,r.length-c|0)}catch(A){o=""}var l=H.d(128);q(i,l),a={hd:l,tl:e},t=o}}(t,0);return M.f(X.c((function(t,a){return{hd:a,tl:t}}),0,a))}function $(t,a){try{return H.f(t,a)}catch(e){return}}var Z=H.f;function tt(t,a,e){var n=e-a|0;return H.g(t,a,n)}var at=e(19),et=e(10),nt=e(16);U.b("Tree-PhyloLibWeb.UnknownClade");function ct(t){return t.TAG?0:{hd:t.clade_id,tl:X.b(X.g(ct,t.children))}}function rt(t){return function(t){var a=ct(t);return et.a(X.l(nt.f,a),X.k(nt.f,a))}(t)?t:L.f("clade_ids not unique")}var gt={contents:0},ot={TAG:0,clade_id:(gt.contents=gt.contents+1|0,gt.contents),children:0,bootstrap:void 0,rank:void 0,id:void 0,name:void 0};function it(t){return!t.TAG&&0===t.children&&void 0===t.bootstrap&&void 0===t.rank&&void 0===t.id&&void 0===t.name}function lt(t,a,e){return{TAG:1,scientific_name:t,id:a,name:e}}function st(t){return{TAG:1,scientific_name:t,id:void 0,name:void 0}}function At(t){return ot.TAG?L.f("Representation invariant broken"):rt({TAG:0,clade_id:(gt.contents=gt.contents+1|0,gt.contents),children:t,bootstrap:void 0,rank:void 0,id:void 0,name:void 0})}var ut={contents:!0},Gt={contents:""};function mt(t){return ut.contents?L.p(t):void(Gt.contents=Gt.contents+at.b(t))}function Tt(t){return ut.contents?L.s(t):void(Gt.contents=Gt.contents+t)}function ft(t){return ut.contents?void console.log(t):Tt(t+"\n")}function ht(t){for(var a=1;a<=t;++a)mt(32)}function dt(t,a,e){for(;;){var n=a,c=t;if(!c)return;var r=c.tl,g=c.hd;ht(g===n&&0!==g?1:((g-n|0)<<1)+(0===g||0===n?0:1)|0),0===r?Tt(e):mt(124),a=g+1|0,t=r}}function Ct(t){return dt(X.i(t),0,"|"),ft("")}function Et(t){return t?dt(X.i(t),0,"o-"):L.i("ds")}function pt(t,a,e){for(;;){var n=t;if(!n)return;var c=n.tl,r=n.hd;if(r.TAG){0!==e?Ct(e):L.r(void 0),0!==e&&Et(e);var g=r.name;void 0!==g&&"Unnamed"===r.scientific_name?ft(g):ft(r.scientific_name),t=c}else{var o=0!==c?{hd:a,tl:e}:e?{hd:a,tl:e.tl}:{hd:a,tl:0};0!==e&&Ct(e),0!==e&&Et(e),Tt("C\n"),pt(r.children,a+1|0,o),t=c}}}function vt(t){return Gt.contents="",ut.contents=!1,pt({hd:t,tl:0},0,0),Gt.contents}var bt=e(18),yt=e(2),wt=e(15);function _t(t,a,e){var n=$(yt.c(e,t),a);return void 0!==n?n:L.f("Index out of bounds")}function Nt(t,a,e){return et.b(et.b(t,a),e)}function It(t,a,e,n,c,r,g,o){return[n.contents<1?-1e9:o+yt.c(yt.c(c,e.contents),n.contents-1|0)|0,e.contents<1?-1e9:o+yt.c(yt.c(c,e.contents-1|0),n.contents)|0,e.contents<1||n.contents<1?-1e9:yt.c(yt.c(c,e.contents-1|0),n.contents-1|0)+(et.a($(t,(0>e.contents?0:e.contents)-1|0),$(a,(0>n.contents?0:n.contents)-1|0))?r:g)|0]}function kt(t,a,e,n,c,r,g){for(var o=function(t,a,e,n,c){for(var r=M.d(n,c,-1e9),g=0;g<n;++g)yt.d(yt.c(r,g),0,Math.imul(g,e));for(var o=0;o<c;++o)yt.d(yt.c(r,0),o,Math.imul(o,e));return r}(0,0,c,r,g),i=1;i<r;++i)for(var l=1;l<g;++l){var s=It(t,a,{contents:i},{contents:l},o,e,n,c);yt.d(yt.c(o,i),l,Nt(s[0],s[1],s[2]))}return o}function St(t,a,e,n,c,r){for(var g={contents:Y(t)},o={contents:Y(a)},i="",l="";g.contents>0||o.contents>0;){var s=It(t,a,g,o,e,n,c,r),A=yt.c(yt.c(e,g.contents),o.contents);A===s[2]?(i=at.b(Z(t,g.contents-1|0))+i,l=at.b(Z(a,o.contents-1|0))+l,g.contents=g.contents-1|0,o.contents=o.contents-1|0):A===s[0]?(i="_"+i,l=at.b(Z(a,o.contents-1|0))+l,o.contents=o.contents-1|0):A===s[1]?(l="_"+l,i=at.b(Z(t,g.contents-1|0))+i,g.contents=g.contents-1|0):L.f("This should not happen")}return[K(i),K(l)]}function xt(t,a,e,n,c){for(var r=St(t,a,kt(t,a,e,n,c,Y(t)+1|0,Y(a)+1|0),e,n,c),g=yt.c(r,0),o=yt.c(r,1),i=Y(g),l=0,s=0;s<i;++s)et.a($(g,s),95)||et.a($(o,s),95)?l=l-c|0:et.d($(g,s),$(o,s))&&(l=l+1|0);return l}var Ot={contents:!0},Rt={contents:""};function Dt(t){return Ot.contents?L.p(t):void(Rt.contents=Rt.contents+at.b(t))}function jt(t){Ot.contents?console.log(t):Rt.contents=Rt.contents+(t+"\n")}function Pt(t,a){for(var e=Y(t),n=0,c=(e-1|0)/80|0;n<=c;++n){jt(tt(t,Math.imul(80,n),nt.h(e,Math.imul(80,n+1|0))));for(var r=Math.imul(80,n),g=nt.h(e-1|0,Math.imul(80,n+1|0)-1|0);r<=g;++r)et.a($(t,r),$(a,r))?Dt(42):et.a($(t,r),95)||et.a($(a,r),95)?Dt(32):Dt(124);jt(""),jt(tt(a,Math.imul(80,n),nt.h(e,Math.imul(80,n+1|0)))),jt("")}}function Xt(t,a,e,n){for(var c=t.length,r=bt.b(void 0,c),g=0;g<c;++g)for(var o=g+1|0;o<c;++o)bt.a(r,[g,o],xt(yt.c(t,g),yt.c(t,o),a,e,n));return r}function Mt(t,a){var e=function(t){return t.length}(t);L.q(e);var n=function(t){return Y(yt.c(t,0))}(t);L.q(n);for(var c=bt.b(void 0,e),r=0;r<e;++r)for(var g=0,o=r+1|0;o<e;++o){for(var i=0;i<n;++i)L.q(r),95===_t(r,i,t)||95===_t(o,i,t)?g=g+a|0:_t(r,i,t)!==_t(o,i,t)&&(g=g+1|0);bt.a(c,[r,o],g)}return c}function Ht(t){if(0===bt.f(t))return[[0,0],0];var a=[[0,0],L.j];return bt.e((function(t,a,e){return a<e[1]?[t,a]:e}),t,a)}function Bt(t){return Ht(t)[0]}function Ft(t,a,e){return bt.d(e,[et.c(t,a),et.b(t,a)])}function Lt(t,a,e){return t>=a?L.f("invalid input"):function(t,a,e){bt.c((function(a,n){return a[0]===t?(n+Ft(a[1],t,e))/2:a[1]!==t?n:(n+Ft(a[0],t,e))/2}),e)}(t,0,e),function(t,a){bt.c((function(a,e){return et.a(a[0],t)||et.a(a[1],t)?void 0:wt.a(e)}),a)}(a,e),e}function Ut(t){return 0===bt.f(t)}function zt(t,a,e,n){var c=yt.c(t,a),r=yt.c(t,e),g=X.a(a,n),o=X.a(e,n);if(void 0!==g){if(void 0!==o){var i=X.h(e,X.h(a,n));return{hd:[a<e?a:e,At({hd:g,tl:{hd:o,tl:0}})],tl:i}}var l=X.h(a,n);return{hd:[a<e?a:e,At({hd:st(r),tl:{hd:g,tl:0}})],tl:l}}if(void 0===o)return{hd:[a<e?a:e,At({hd:st(c),tl:{hd:st(r),tl:0}})],tl:n};var s=X.h(e,n);return{hd:[a<e?a:e,At({hd:st(c),tl:{hd:o,tl:0}})],tl:s}}function Vt(t,a){var e=a.length;if(0===e)return ot;if(1===e)return st(yt.c(a,0));for(var n=t,c=0;;){var r=c,g=n;if(Ut(g)){var o=X.j((function(t){return t[1]}),r);return 1===X.f(o)?X.e(o):X.f(o)>1?At(o):L.f("Precondition violated in upgma_help")}var i=Bt(g),l=i[1],s=i[0];c=zt(a,s,l,r),n=Lt(s,l,g)}}var qt=e(1),Wt=e(138),Jt={contents:!0},Kt={contents:""},Yt={contents:L.v};function Qt(t){return Jt.contents?qt.a(Wt.a(Yt.contents,{_0:{TAG:2,_0:0,_1:0},_1:"%s"}),t):void(Kt.contents=Kt.contents+t)}function $t(t){for(var a=1;a<=t;++a)Qt("  ")}function Zt(t,a,e,n){return $t(e),Qt("<"+t),function(t){for(;;){var a=t;if(!a)return;var e=a.tl,n=a.hd,c=0!==e;Qt(n[0]+'="'+n[1]+'"'+(c?" ":"")),t=e}}(a),Qt(n?">\n":">")}function ta(t,a){return $t(a),Qt("</"+t+">\n")}function aa(t,a,e){return $t(a),Qt("<"+t+">"),Qt(e),ta(t,0)}function ea(t,a,e){if(void 0!==e)return aa(t,a,e)}function na(t){return Zt("phyloxml",0,0,!0),Zt("phylogeny",0,1,!0),0!==t.name.length&&aa("name",2,t.name),0!==t.description.length&&aa("description",2,t.description),it(t.tree)||function t(a,e){if(a.TAG)return Zt("clade",0,e,!0),ea("name",e+1|0,a.name),""===a.scientific_name&&void 0===a.id||(Zt("taxonomy",0,e+1|0,!0),""!==a.scientific_name&&aa("scientific_name",e+2|0,a.scientific_name),ea("id",e+2|0,a.id),ta("taxonomy",e+1|0)),ta("clade",e);Zt("clade",0,e,!0),ea("name",e+1|0,a.name);var n=a.bootstrap;return ea("confidence",e+1|0,void 0!==n?L.x(n):void 0),void 0===a.rank&&void 0===a.id||(Zt("taxonomy",0,e+1|0,!0),ea("id",e+2|0,a.id),ea("rank",e+2|0,a.rank),ta("taxonomy",e+1|0)),function(a,e){for(;;){var n=a;if(!n)return;t(n.hd,e),a=n.tl}}(a.children,e+1|0),ta("clade",e)}(t.tree,2),ta("phylogeny",1),ta("phyloxml",0)}var ca=m.a.Content;function ra(){var t=Object(n.useState)(""),a=Object(G.a)(t,2),e=a[0],r=a[1],g=Object(n.useState)(!1),o=Object(G.a)(g,2),i=o[0],l=o[1],A=Object(n.useState)(!0),m=Object(G.a)(A,2),d=m[0],E=m[1],w=Object(n.useState)(!0),_=Object(G.a)(w,2),N=(_[0],_[1]),X=Object(n.useState)([]),M=Object(G.a)(X,2),H=M[0],B=M[1],F=Object(n.useState)([]),L=Object(G.a)(F,2),U=L[0],z=L[1],V=Object(n.useState)(void 0),q=Object(G.a)(V,2),J=q[0],Y=q[1],$=Object(n.useState)(!1),Z=Object(G.a)($,2),tt=Z[0],at=Z[1],et=Object(n.useState)(!1),nt=Object(G.a)(et,2),ct=nt[0],rt=nt[1],gt=[">H1N1 PB-2 Genome\nATGGAGAGAATAAAAGAACTAAGAGATTTGATGTCGCAGTCTCGCACTCGCGAGATACTGACAAAAACCACTGTGGACCA\nTATGGCCATAATCAAGAAATATACGTCAGGAAGACAGGAGAAGAATCCTGCTCTTAGGATGAAATGGATGATGGCAATGA\nAATACCCGATTACAGCAGACAAAAGGATAATGGAGATGATCCCTGAGAGAAATGAGCAAGGTCAGACTCTTTGGAGCAAA\nACAAATGATGCTGGATCAGACAGAGTGATGGTATCACCTCTGGCTGTGACGTGGTGGAATAGAAATGGGCCTACGACAAG\nTACAGTCCATTATCCAAAGGTCTATAAAACCTACTTTGAAAAAGTCGAAAGGTTAAAACATGGAACCTTCGGTCCAGTTC\nACTTTCGAAATCAAGTTAAAATACGCCGCAGGGTTGACATAAACCCGGGCCATGCAGATCTCAGTGCTAAAGAAGCACAA\nGATGTCATCATGGAGGTCGTTTTCCCAAATGAGGTCGGAGCCAGGATATTGACATCAGAGTCACAGTTAACAATAACAAA\nGGAAAAGAAGGAGGAGCTTCAGGACTGTAAGATTGCTCCTTTAATGGTGGCATACATGTTGGAGAGAGAACTGGTTCGCA\nAAACCAGATTCCTACCAGTAGCTGGCGGGACAAGCAGCGTGTATATCGAGGTATTACACTTGACTCAAGGGACCTGCTGG\nGAACAAATGTACACACCGGGAGGGGAGGTGAGAAATGATGATGTTGATCAGAGTTTAATTATTGCTGCCAGAAATATTGT\nTAGGAGAGCGACAGTATCAGCAGACCCGTTGGCTTCGCTCTTGGAGATGTGCCATAGTACACAAATTGGCGGGATAAGAA\nTGGTAGACATCCTTAGACAAAACCCAACAGAAGAGCAAGCTGTGGATATATGCAAAGCAGCAATGGGTCTAAGGATCAGT\nTCATCCTTCAGTTTTGGAGGTTTCACTTTCAAAAGGACAAGTGGGTCATCTGTCAAAAGAGAAGAAGAAGTGCTCACAGG\nCAACCTCCAAACATTGAAAATAAGAGTGCATGAAGGATATGAGGAATTCACAATGGTTGGGCGAAGAGCGACAGCCATTC\nTAAGGAAAGCAACCAGAAGGCTGATCCAACTGATAGTGAGTGGGAGAGACGAGCAGTCAATCGCCGAAGCAATCATAGTG\nGCAATGGTGTTCTCACAAGAGGATTGTATGATAAAAGCAGTACGAGGTGATTTGAATTTTGTCAATAGGGCGAACCAGCG\nGCTAAACCCTATGCATCAACTCCTGAGACATTTCCAAAAGGATGCAAAGGTGCTGTTTCAAAACTGGGGGATTGAACCCA\nTTGACAATGTCATGGGGATGATCGGAATATTACCTGACATGACCCCCAGCACAGAGATGTCACTAAGAGGAGTGAGAGTC\nAGCAAAACGGGAGTGGATGAATATTCCAGCACTGAGAGAGTGGTCGTGAGCATTGATCGTTTCTTGAGGGTCCGAGACCA\nGAGGGGAAACGTGCTCTTGTCTCCTGAAGAAGTTAGTGAAACACAGGGAACAGAAAAGCTGACGATAACATATTCATCGT\nCCATGATGTGGGAAATCAATGGCCCGGAATCAGTGCTAGTTAACACATATCAATGGATCATTAGAAATTGGGAGACTGTG\nAAAATTCAGTGGTCCCAAGACCCTACAATGCTATACAATAAGATGGAATTTGAGCCCTTTCAATCCTTGGTGCCTAAGGC\nTGCCAGAGGTCAGTATAGTGGATTTGTGAGGACGCTATTCCAGCAGATGCGTGATGTGCTGGGGACATTTGACACTGTCC\nAGATAATAAAGCTACTACCATTTGCAGCAGCCCCACCGGAACAGAGTAGGATGCAGTTCTCTTCTCTAACTGTGAACGTA\nAGGGGTTCAGGAATGAGAATACTTGTGAGAGGCAACTCCCCTGTGTTCAACTATAACAAGGCAACCAAGAGGCTCACAGT\nCCTTGGAAAGGATGCAGGTGCATTGACAGAAGATCCAGATGAGGGAACAGCAGGAGTGGAATCCGCGGTATTAAGAGGAT\nTTCTAATTCTGGGCAAAGAAGACAAAAGATATGGACCAGCGTTGAGCATCAACGAATTGAGCAATCTTGCAAAAGGGGAG\nAAGGCTAACGTGTTGATAGGGCAAGGAGACGTGGTGTTGGTAATGAAACGGAAACGGGACTCTAGCATACTTACTGACAG\nCCAGACAGCGACCAAAAGAATTCGGATGGCCATCAATTAG ","\n>H3N2 PB-2 Genome\nagcgaaagcaggtcaaatatattcaatatggagagaataaaagaactaagagatttgatgtcgcagtctcgcactcgcga\ngatactgacaaaaaccactgtggaccatatggccataatcaagaaatatacgtcaggaagacaggagaagaaccctgctc\nttaggatgaagtggatgatggcaatgaaatatccgattacagcagacaggaggataatggagatgatccctgaaagaaat\ngagcaaggtcagactctttggagcaaaacgaatgatgctggatcagacagagtgatggtatcgcctctggctgtgacatg\ngtggaatagaaatggaccaacgacaagtacagtccattatccgaaggtctacaaaacctactttgagaaggtcgaaaggt\ntaaagcatggaacctttggtcccgttcactttcggaatcaagttaaaatacgccgcagggttgacataaacccgggccac\ngcagacctcagtgctaaagaagcacaagatgtcatcatggaggtcgttttcccaaacgaagtaggagccaggatattaac\natcagagtcacaattaacaataacaaaagaaaaaaaggaagagctccaagactgtaagattgctcctttgatggtggcat\nacatgttggaaagagaattagttcgcaaaaccagattcctaccggtagctggtgggacaagcagcgtgtatatcgaggtg\nttgcacttgacccaggggacctgccgggaacaaatgtatacaccgggaggagaggtgagaaatgatgatattgatcagag\ntttaattattgctgctagaaatattgtcagaagagcaacagtatcagcagatccattagcttcgctcttggaaatgtgcc\natagcacacaaattggtggaataaggatggtagatattcttaggcaaaatccaacagaagagcaagctgtggatatatgc\naaagcagcaatgggtctgaggatcagttcatccttcagctttggaggtttcactttcaaaaggacaagtggatcatctgt\ncaaaagagaagaagaagtactcacaggcaaccttcagacattgaaaataagagtgcacgaaggatatgaggaatttacaa\ntggtagggcgaagagcaacagccattttaaggaaagcaaccagaaggctgattcaattaatagtgagtggaagagacgag\ncaatcaattgccgaagcaatcatagtggcaatggtgttctcacaagaggattgcatgataaaggcagtccgaggcgattt\ngaattttgtcaacagagcgaaccagcggctgaatcctatgcatcaactcctgaggcacttccagaaggatgcaaaggtgc\ntatttcaaaactggggagttgaacccattgacaatgtcatggggatgatcggaatattacctgacatgacccccagcaca\ngaaatgtcactaaggggagtaagggtcagtaaaatgggagtggatgaatattccagtactgagagagtggtgtgagcat\ntgaccgtttcttgagggtccgagaccagaggggaaatgtgctcttgtctcctgaagaagttagtgaaacacagggaacag\naaaagctgacaataacatattcatcgtccatgatgtgggaaatcaatggcccggaatcagtgctggtcaacacatatcag\ntggatcattagaaactgggaaactgtgaagatccagtggtcccaagaccctacaatgctatacaataagatgaaatttga\ngccttttcaatctttagtgcctaaggctgccagaggccagtatagcggctttgtgaggacgctatttcagcaaatgcgtg\natgtgctgggaacatttgacactgtccagataataaagctactaccatttgcagcagccccgccagaacagagtagaatg\ncagttctcttctctaactgtgaacgtaaggggttcaggaatgagaatacttgtgagaggtaattctcctgtgttcaatta\ntaacaaggcaaccaagaggctcacaatccttggaaaggacgcaggtgcactgactgaagatccagatgaaggaacagcag\ngagtggagtctgcagtattaagaggatttctaattctgggcaaagaagacaaaagatatggaccagcattgagcatcaac\ngaattgagcaatcttgcgaaaggggagaaggctaatgtgttgatagggcaaggagacgtggtgttggtaatgaaacggaa\nacgggactctagcatacttactgacagccagacagcgaccaaaagaattcggatggccatcaattagtgtcgaattgttt\naaaaacgaccttgtttctact\n","\n>H5N1 PB-2 Genome\natggagagaataaaagaattacgagatctaatgtcacagtcccgcactcgcgagatactaacaaaaaccactgtggacca\ntatggccataatcaagaaatacacatcaggaagacaagagaagaaccctgctctcagaatgaaatggatgatggcaatga\naatatccaatcacagcggacaagagaataatagagatgattcctgaaaggaatgaacaagggcagacgctctggagcaag\nacaaatgatgctggatcggacagggtgatggtgtctcccctagctgtaacttggtggaacaggaatgggccggcgacaag\ntgcagttcattatccaaaggtttacaaaacatactttgagaaggttgaaaggttaaaacatggaacctttggtcccgttc\natttccgaaaccaagttaaaatacgccgccgagttgatataaatcctggccatgcagatctcagtgctaaagaagcacaa\ngatgtcatcatggaggtcgttttcccaaatgaagtgggagccagaatattgacatcagagtcgcaattgacaataacgaa\nagagaagaaagaagagctccaagattgtaagattgctcccttaatggttgcatacatgttggaaagggaactagtccgca\naaaccagattcctaccggtagcaggcggaacaagcagtgtgtacattgaggtattgcatttgactcaagggacctgctgg\ngaacagatgtacactccaggcggagaagtgagaaatgacgatgttgaccagagtttgatcatcgctgccagaaacattgt\ntaggagagcaacggtatcagcggatccactggcatcactgctggagatgtgtcacagcacacaaattggtgggataagga\ntggtggacatccttaggcaaaatccaactgaggaacaagctgtggatatatgcaaagcagcaataggtctgaggatcagt\ntcttcctttagctttggaggcttcactttcaaaagaacaagtggatcatccgtcaagaaggaagaggaagtgcttacagg\ncaacctccaaacattgaaaataagagtacatgaggggtatgaggaattcacaatggttgggcggagggcaacagctatcc\ntgaggaaagcaactagaaggctgattcagttgatagtaagtggaagagacgaacaatcaatcgctgaggcaatcattgta\ngcaatggtgttctcacaggaggattgcatgataaaggcagtccgaggcgatctgaatttcgtaaacagagcaaaccaaag\nattaaaccccatgcatcaactcctgagacattttcaaaaggatgcaaaagtgctatttcagaattggggaattgaaccca\nttgataatgtcatggggatgatcggaatattacctgacatgactcccagcacagaaatgtcactgagaggagtaagagtt\nagtaaaatgggagtggatgaatattccagcactgagagagtagttgtaagtattgaccgtttcttaagggttcgagatca\ngcgggggaacgtactcttgtctcccgaagaggtcagcgaaacccagggaacagagaaattgacaataacatattcatcat\ncaatgatgtgggaaatcaacggtcctgagtcagtgcttgttaacacttatcagtggatcatcagaaactgggagactgtg\naagattcaatggtctcaagaccccacgatgctgtacaataagatggagtttgaaccgttccaatccttggtacccaaggc\ntgccagaggtcaatacagtggatttgtgagaacattattccaacaaatgcgtgacgtactggggacatttgatactgtcc\nagataataaagctgctaccatttgcagcagccccaccgaagcagagcagaatgcagttttcttctttaactgtgaatgtg\nagaggctcaggaatgagaatactcgtaaggggcaattcccctgtattcaactataataaggcaaccaaaaggcttaccgt\ntcttggaaaggacgcaggtgcattaacagaggatccagatgaggggacagccggagtggaatctgcagtactgaggggat\ntcttaattctaggcaaggaggacaaaaggtatgggccagcattgagcatcaatgaactgagcaatcttgcgaagggggag\naaagctaatgtgctgatagggcaaggagacgtggtgttggtaatgaaacggaaacgggactctagcatacttactgacag\nccagacagcgaccaaaagaattcggatggccatcaattag "],ot=["H1N1","H3N2","H5N1"],it=Object(n.useRef)(null),lt=c.a.createElement(D.a,{style:{fontSize:24},spin:!0}),st=function(t,a){B((function(a){return a.concat(t)})),z((function(t){return t.concat(a)}))},At=function(){var t=Object(k.a)(I.a.mark((function t(a,e){var n;return I.a.wrap((function(t){for(;;)switch(t.prev=t.next){case 0:try{(n=new FileReader).onload=function(){var t=K(n.result);st(t,e)},n.readAsText(a)}catch(c){console.log(c),console.log("File printing failed")}case 1:case"end":return t.stop()}}),t)})));return function(a,e){return t.apply(this,arguments)}}(),ut=function(){window.scrollTo({behavior:"smooth",top:it.current.offsetTop-15})},Gt=function(){null!==it.current?ut():setTimeout((function(){ut()}),1e3)},mt=function(t){return t.split(".").slice(0,-1).join(".").toUpperCase()},Tt={accept:".FASTA, .txt, .fasta",action:"https://www.mocky.io/v2/5cc8019d300000980a055e76",headers:{authorization:"authorization-text"},multiple:!0,transformFile:function(t){at(!0);var a=mt(t.name);At(t,a)},onRemove:function(t){var a=mt(t.name);console.log(a);var e=U.indexOf(a);console.log(e),H.splice(e,1),U.splice(e,1),B(H),z(U)}};return c.a.createElement("div",{className:"wrapper"},c.a.createElement(ca,{justify:"center"},c.a.createElement(T.a,{className:"page",justify:"center"},c.a.createElement("div",null,c.a.createElement("h1",null,"Generate a Phylogenetic Tree"),c.a.createElement("h2",null,"By computing similarity scores for DNA samples of species, we can infer their species' evolutionary history through time."),c.a.createElement("h2",null,"Begin by uploading"," ",c.a.createElement(p,{content:c.a.createElement("p",null,"The FASTA format is a way of representing nucleotide sequences. It was popularized by the FASTA software developed by David J. Lipman and William R. Pearson."),vocab:".FASTA",link:"https://www.ncbi.nlm.nih.gov/pmc/articles/PMC280013/",linkText:"the original paper on the FASTA program"})," ","files that contain DNA sequences, or use our example DNA sequences."))),c.a.createElement(T.a,{className:"centered-content"},c.a.createElement("h3",{className:"example-text"},"See some examples:")),c.a.createElement(T.a,{className:"centered-content",gutter:[5,5]},c.a.createElement(f.a,{lg:1}),c.a.createElement(f.a,{lg:10,md:12,className:"centered-content"},c.a.createElement(C.default.Group,{onChange:function(t){switch(t.target.value){case"Influenza A Viruses":l(!0),r(v);break;case"COXII":l(!0),r(b);break;case"Virus Capsid":l(!0),console.log(y),r(y)}}},c.a.createElement(C.default.Button,{value:"Virus Capsid"},"Virus Capsid Gene"),c.a.createElement(C.default.Button,{value:"COXII"},"Cytochrome C Oxidase Subunit II"),c.a.createElement(C.default.Button,{value:"Influenza A Viruses"},"Influenza A PB-2"))),c.a.createElement(f.a,{lg:1,md:12},c.a.createElement(h.a,null,c.a.createElement("a",{href:"/phylo-lib-web/examples/FASTA/phylo_examples.zip"},"Download Examples")))),c.a.createElement(T.a,{className:"centered-content"},c.a.createElement("h3",{className:"upload-text"},"Or upload 3 or more FASTA files:")),c.a.createElement(T.a,{className:"horizontally-centered"},c.a.createElement(s.a,{content:c.a.createElement("div",null,c.a.createElement("p",null,"Clustal and DP are two techniques of aligning DNA sequences. Clustal is a software package used for multiple sequence alignment, utilizing heuristics for faster alignments. The dynamic programming (DP) technique generates a globally optimum alignment, but may be situationally slower than Clustal."),null),title:"MSA Information",trigger:"click"},c.a.createElement("div",{className:"generate-info"},c.a.createElement(u.a,null))),c.a.createElement("div",{style:{display:"flex",flexDirection:"column"}},c.a.createElement(S.a,{checkedChildren:"Clustal",unCheckedChildren:"DP",className:"generate-toggle",defaultChecked:!0,onChange:function(t,a){E(t),t||N(!0)}}),null),c.a.createElement(x.a,Tt,c.a.createElement(h.a,null,c.a.createElement(j.a,null)," Upload .FASTA files")),c.a.createElement(h.a,{onClick:function(){rt(!0);var t=tt?U:ot;if(d){var a="",e=new URLSearchParams;e.append("email","vg222@cornell.edu");for(var n="",c=0;c<H.length;c++)n+=">",n+=U[c].split(" ").join("_"),n+="\n",n+=W(H[c]),n+="\n";e.append("sequence",n),e.append("outfmt","fa"),fetch("https://www.ebi.ac.uk/Tools/services/rest/clustalo/run",{method:"POST",headers:{"Content-Type":"application/x-www-form-urlencoded"},body:e,redirect:"follow"}).then((function(t){return t.text()})).then((function(e){a=e,console.log(a),function e(n){fetch("https://www.ebi.ac.uk/Tools/services/rest/clustalo/status/"+a,{method:"GET",redirect:"follow"}).then((function(t){return t.text()})).then((function(c){if("FINISHED"!==c&&n<50)console.log("querying"),setTimeout(e(n+1),1e3);else{if("FINISHED"!==c)return;console.log(c),fetch("https://www.ebi.ac.uk/Tools/services/rest/clustalo/result/"+a+"/aln-fasta",{method:"GET",redirect:"follow"}).then((function(t){return t.text()})).then((function(a){console.log("got result"),console.log(a);var e=Q(a);console.log("got dnas");var n=e;console.log("got msa");var c=Mt(n,1);console.log("got dist mat");var g=vt(Vt(c,t));r(g),l(!0),rt(!1),Gt()})).catch((function(t){return console.log("error1",t)}))}})).catch((function(t){return console.log("error",t)}))}(0)})).catch((function(t){return console.log("error",t)}))}else setTimeout((function(){var a=vt(Vt(Xt(tt?H:gt,1,-1,-1),t));r(a),l(!0),rt(!1),Gt()}),500)},className:"action-button",disabled:H.length<3},"Generate tree"),ct?c.a.createElement(O.a,{className:"spinner",indicator:lt}):null),H.length<3?null:c.a.createElement(T.a,{className:"centered-content"},c.a.createElement(h.a,{onClick:function(){if(tt&&H.length<1)alert("At least one FASTA file must be uploaded for tree generation.");else{var t=tt?U:ot,a=function(t){return Jt.contents=!1,Kt.contents="",na({name:"",description:"",tree:t}),Kt.contents}(Vt(Xt(tt?H:gt,1,-1,-1),t)),e=c.a.createElement("a",{href:"data:text/xml;charset=utf-8,"+encodeURIComponent(a),download:"tree.xml"},"Click to Download");Y(e)}}},"Save tree as phyloXML")),void 0!==J?c.a.createElement(T.a,{className:"centered-content"},J,c.a.createElement(h.a,{onClick:function(){return Y(void 0)},className:"hide-download"},c.a.createElement(R.a,{title:"Hide Download Link"},c.a.createElement(P.a,{style:{color:"firebrick"}})))):null),i?c.a.createElement(T.a,{justify:"center"},c.a.createElement("div",{className:"ascii-phylo-container",ref:it},c.a.createElement("p",{className:"ascii-phylo"},e))):null)}var ga=e(17);function oa(t){return"number"!==typeof t||(t>=14?t<17:t<9)}function ia(t){if("number"!==typeof t)return t.TAG?t._0:String(t._0);switch(t){case 0:return"phylogeny";case 1:return"name";case 2:return"description";case 3:return"clade";case 4:return"rank";case 5:return"confidence";case 6:return"taxonomy";case 7:return"scientific_name";case 8:return"id";case 9:return"<";case 10:return"</";case 11:return">";case 12:return"quote";case 13:return"=";case 14:return".";case 15:return"true";case 16:return"false";case 17:return"EOF";case 18:return"Unit"}}var la=bt.b(void 0,16);function sa(t,a){var e;if(a){var n=L.k(t);e=B.b((function(t){try{return L.h(n)}catch(e){var a=z.a(e);if("End_of_file"===a.RE_EXN_ID)return;throw a}}))}else{var c=B.f(t);e=B.b((function(t){try{return function(t,a){for(;;){var e=a,n=B.g(t);if(void 0===n){if(""===e)throw{RE_EXN_ID:"End_of_file",Error:new Error};return e}if(B.d(t),10===n)return e;if(13===n)return e;a=e+at.b(n)}}(c,"")}catch(e){var a=z.a(e);if("End_of_file"===a.RE_EXN_ID)return;throw a}}))}var r,g=B.g(e);if(void 0===g)return e;try{r=F.h(g,0,5)}catch(i){var o=z.a(i);if("Invalid_argument"===o.RE_EXN_ID)return e;throw o}return"<?xml"!==r||B.d(e),e}function Aa(t,a){var e=B.g(t);if(void 0!==e){var n;try{n=e}catch(r){var c=z.a(r);if(c.RE_EXN_ID===B.a)return ua(a);throw c}return 32!==n?bt.g(la,a)?bt.d(la,a):function(t){return t>=60?t<63:34===t}(e)?{TAG:1,_0:a}:(B.d(t),Aa(t,a+at.b(e))):ua(a)}return ua(a)}function ua(t){return bt.g(la,t)?bt.d(la,t):{TAG:1,_0:t}}function Ga(t){return!(t>57||t<48)}function ma(t,a){for(;;){var e=a,n=B.g(t);if(void 0===n)return{TAG:0,_0:ga.i(e)};if(!Ga(n))return{TAG:0,_0:ga.i(e)};B.d(t),a=e+at.b(n)}}function Ta(t){var a;try{a=function(t){var a;try{a=B.e(t)}catch(n){var e=z.a(n);if(e.RE_EXN_ID===B.a)throw{RE_EXN_ID:"End_of_file",Error:new Error};throw e}return B.f(a)}(t)}catch(s){var e=z.a(s);if("End_of_file"===e.RE_EXN_ID)return{hd:17,tl:0};throw e}for(var n=0;;){var c,r=n,g=0;try{c=B.e(a),g=1}catch(A){var o=z.a(A);if(o.RE_EXN_ID===B.a)return X.i(r);if("End_of_file"===o.RE_EXN_ID)return{hd:17,tl:0};throw o}if(1===g){var i=0;if(c>=32){if(46===c){n={hd:14,tl:r};continue}if(c>=60)if(c>=63)i=2;else switch(c-60|0){case 0:var l=B.g(a);if(void 0===l)return X.i({hd:9,tl:r});if(47===l){B.d(a),n={hd:10,tl:r};continue}n={hd:9,tl:r};continue;case 1:n={hd:13,tl:r};continue;case 2:n={hd:11,tl:r};continue}else if(c>=35)i=2;else switch(c-32|0){case 0:continue;case 1:i=2;break;case 2:n={hd:12,tl:r};continue}}else if(c>=11){if(13===c)continue;i=2}else{if(c>=9)continue;i=2}if(2===i){if(Ga(c)){n={hd:ma(a,at.b(c)),tl:r};continue}n={hd:Aa(a,at.b(c)),tl:r};continue}}}}bt.a(la,"phylogeny",0),bt.a(la,"name",1),bt.a(la,"description",2),bt.a(la,"clade",3),bt.a(la,"rank",4),bt.a(la,"confidence",5),bt.a(la,"taxonomy",6),bt.a(la,"scientific_name",7),bt.a(la,"id",8),bt.a(la,"true",15),bt.a(la,"false",16);var fa=U.b("Phylo_parser-PhyloLibWeb.SyntaxError"),ha={name:"",description:"",tree:ot},da={name:void 0,rank:void 0,confidence:void 0,taxonomy:void 0};var Ca={contents:function(t){return 17}},Ea={contents:function(t){return 17}};function pa(t){var a=qt.a(Ca.contents,void 0);if(!et.a(a,t))throw ia(t),ia(a),{RE_EXN_ID:fa,Error:new Error};qt.a(Ea.contents,void 0)}function va(t){for(;;){var a=t,e=qt.a(Ca.contents,void 0);if("number"===typeof e){switch(e){case 1:case 2:case 3:case 4:case 5:case 6:case 8:case 14:case 15:break;default:return a}pa(e),t=""===a?ia(e):a+" "+ia(e)}else{if(e.TAG){var n=e._0;if(pa({TAG:1,_0:n}),""!==a){t=a+" "+n;continue}t=n;continue}var c=e._0;if(pa({TAG:0,_0:c}),""!==a){t=a+" "+String(c);continue}t=String(c)}}}function ba(t,a){return void 0!==t?{hd:a,tl:t}:{hd:a,tl:0}}function ya(t){return pa(10),pa(t),pa(11)}function wa(t,a){var e=qt.a(Ca.contents,void 0);if("number"===typeof e){if(12===e){pa(12);var n,c=qt.a(Ca.contents,void 0);if("number"===typeof c)if(15!==c){if(16!==c)throw{RE_EXN_ID:fa,Error:new Error};pa(16),n={tag_name:t.tag_name,str_attr:t.str_attr,num_attr:t.num_attr,bool_attr:ba(t.bool_attr,[a,!1])}}else pa(15),n={tag_name:t.tag_name,str_attr:t.str_attr,num_attr:t.num_attr,bool_attr:ba(t.bool_attr,[a,!0])};else{var r=va("");n={tag_name:t.tag_name,str_attr:ba(t.str_attr,[a,r]),num_attr:t.num_attr,bool_attr:t.bool_attr}}return pa(12),n}throw{RE_EXN_ID:fa,Error:new Error}}if(e.TAG)throw{RE_EXN_ID:fa,Error:new Error};var g=e._0;return pa({TAG:0,_0:g}),{tag_name:t.tag_name,str_attr:t.str_attr,num_attr:ba(t.num_attr,[a,g]),bool_attr:t.bool_attr}}function _a(t){pa(9);var a=qt.a(Ca.contents,void 0);if(!function(t){if("number"!==typeof t)return 1===t.TAG;switch(t){case 0:case 1:case 2:case 3:case 4:case 5:case 6:case 7:case 8:return!0;default:return!1}}(a))throw{RE_EXN_ID:fa,Error:new Error};pa(a);for(var e={tag_name:a,str_attr:void 0,num_attr:void 0,bool_attr:void 0};;){var n=e,c=qt.a(Ca.contents,void 0);if(!oa(c)){if("number"===typeof c){if(11!==c)throw{RE_EXN_ID:fa,Error:new Error};return pa(11),n}throw{RE_EXN_ID:fa,Error:new Error}}pa(c),pa(13),e=wa(n,ia(c))}}function Na(t){for(;;){var a=qt.a(Ca.contents,void 0);if("number"!==typeof a)pa(a);else{if(9!==a){if(10===a)return ya(t);pa(a);continue}Na(_a().tag_name)}}}function Ia(t){if("number"===typeof qt.a(Ca.contents,void 0))throw{RE_EXN_ID:fa,Error:new Error};var a=va("");return ya(1),a}function ka(t){if("number"===typeof qt.a(Ca.contents,void 0))throw{RE_EXN_ID:fa,Error:new Error};var a=va("");return ya(2),a}function Sa(t){pa(14);var a=qt.a(Ca.contents,void 0);if("number"===typeof a)return ya(5),0;if(a.TAG)return ya(5),0;var e=a._0,n=ga.a(String(t)+"."+String(e));return pa({TAG:0,_0:e}),ya(5),n}function xa(t){var a=qt.a(Ca.contents,void 0);if("number"===typeof a){if(14===a)return Sa(0);throw{RE_EXN_ID:fa,Error:new Error}}if(a.TAG)throw{RE_EXN_ID:fa,Error:new Error};var e=a._0;return pa({TAG:0,_0:e}),14===qt.a(Ca.contents,void 0)?Sa(e):(ya(5),e)}function Oa(t){if("number"===typeof qt.a(Ca.contents,void 0))throw{RE_EXN_ID:fa,Error:new Error};var a=va("");return ya(8),a}function Ra(t){if("number"===typeof qt.a(Ca.contents,void 0))throw{RE_EXN_ID:fa,Error:new Error};var a=va("");return ya(7),a}function Da(t){for(;;){var a=t,e=qt.a(Ca.contents,void 0);if("number"!==typeof e)throw{RE_EXN_ID:fa,Error:new Error};if(9!==e){if(10!==e)throw{RE_EXN_ID:fa,Error:new Error};return ya(6),a}var n=_a().tag_name;if("number"!==typeof n)Na(n);else{if(7!==n){if(8!==n){Na(n);continue}t={id:Oa(),scientific_name:a.scientific_name};continue}t={id:a.id,scientific_name:Ra()}}}}function ja(t,a){var e=qt.a(Ca.contents,void 0);if("number"===typeof e){if(9!==e){if(10!==e)throw{RE_EXN_ID:fa,Error:new Error};if(ya(3),it(t)){var n=a.taxonomy;if(void 0!==n)return lt(n.scientific_name,n.id,a.name);var c=a.name;return void 0!==c?lt(c,void 0,c):st("Unnamed")}if(t.TAG)return t;var r=a.taxonomy,g=void 0!==r?r.id:void 0;return{TAG:0,clade_id:t.clade_id,children:t.children,bootstrap:a.confidence,rank:a.rank,id:g,name:a.name}}var o=_a().tag_name;if("number"===typeof o)switch(o){case 1:return ja(t,{name:Ia(),rank:a.rank,confidence:a.confidence,taxonomy:a.taxonomy});case 3:if(it(t))return ja(ja(t,da),a);var i=ja(ot,da);return ja(At({hd:t,tl:{hd:i,tl:0}}),a);case 5:return ja(t,{name:a.name,rank:a.rank,confidence:xa(),taxonomy:a.taxonomy});case 6:return ja(t,{name:a.name,rank:a.rank,confidence:a.confidence,taxonomy:Da({id:void 0,scientific_name:"Unnamed"})})}return Na(o),ja(t,a)}throw{RE_EXN_ID:fa,Error:new Error}}function Pa(t){var a=function(t){var a={contents:Ta(t)},e={contents:function(t,a){return 17}};return e.contents=function(n){return n?function(n){var c=a.contents;return c?c.hd:(a.contents=Ta(t),qt.b(e.contents,!0,void 0))}:function(e){var n=a.contents;return n?(a.contents=n.tl,18):(a.contents=Ta(t),18)}},e.contents}(t);Ca.contents=qt.a(a,!0),Ea.contents=qt.a(a,!1);var e=qt.a(Ca.contents,void 0);if("number"===typeof e){if(9===e){var n=_a().tag_name;if("number"===typeof n)throw{RE_EXN_ID:fa,Error:new Error};if(1===n.TAG){if("phyloxml"===n._0){var c=_a().tag_name;if("number"===typeof c){if(0!==c)throw{RE_EXN_ID:fa,Error:new Error};var r=function(t){for(;;){var a=t,e=qt.a(Ca.contents,void 0);if("number"!==typeof e)throw{RE_EXN_ID:fa,Error:new Error};if(9!==e){if(10!==e)throw{RE_EXN_ID:fa,Error:new Error};return ya(0),a}var n=_a().tag_name;if("number"===typeof n)if(n>=4)Na(n);else switch(n){case 0:Na(n);continue;case 1:t={name:Ia(),description:a.description,tree:a.tree};continue;case 2:t={name:a.name,description:ka(),tree:a.tree};continue;case 3:t={name:a.name,description:a.description,tree:ja(a.tree,da)};continue}else Na(n)}}(ha);return ya({TAG:1,_0:"phyloxml"}),r}throw{RE_EXN_ID:fa,Error:new Error}}throw{RE_EXN_ID:fa,Error:new Error}}throw{RE_EXN_ID:fa,Error:new Error}}if(17===e)return ha;throw{RE_EXN_ID:fa,Error:new Error}}throw{RE_EXN_ID:fa,Error:new Error}}function Xa(t){return Pa(function(t){return sa(t,!1)}(t))}var Ma=m.a.Content;function Ha(){var t=Object(n.useState)([]),a=Object(G.a)(t,2),e=a[0],r=a[1],g=Object(n.useState)(void 0),o=Object(G.a)(g,2),i=o[0],l=o[1],A=Object(n.useState)(!1),m=Object(G.a)(A,2),f=m[0],d=m[1],E=Object(n.useState)(""),v=Object(G.a)(E,2),b=v[0],y=v[1],w=Object(n.useState)(""),_=Object(G.a)(w,2),N=_[0],I=_[1],k=Object(n.useRef)(null),S=new FileReader,O=function(t){try{S.onload=function(){var t=vt(Xa(S.result).tree);y(t),d(!0),window.scrollTo({behavior:"smooth",top:k.current.offsetTop-15}),k.current.scrollTo({behavior:"smooth",top:1e3})},S.readAsText(t)}catch(a){console.log(a),console.log("File printing failed")}},R={accept:".xml",action:"https://www.mocky.io/v2/5cc8019d300000980a055e76",fileList:e,headers:{authorization:"authorization-text"},onChange:function(t){r(t.fileList)},transformFile:function(t){r([t]),l(t)}};return c.a.createElement("div",{className:"wrapper"},c.a.createElement(Ma,{justify:"center"},c.a.createElement(T.a,{className:"page",justify:"center",gutter:[16,16]},c.a.createElement("div",null,c.a.createElement("h1",null,"Visualize PhyloXML"),c.a.createElement("h2",null,"Visualize an existing"," ",c.a.createElement(p,{content:c.a.createElement("p",null,"A tree that displays the evolutionary relationships between various organisms."),vocab:"phylogenetic tree",link:"https://www.khanacademy.org/science/high-school-biology/hs-evolution/hs-phylogeny/a/phylogenetic-trees",linkText:"Khan Academy"}),". Begin by uploading a"," ",c.a.createElement(p,{content:c.a.createElement("p",null,"An XML format for describing phylogenetic trees."),vocab:"phyloXML",link:"http://phyloxml.org/",linkText:" the phyloXML website "})," ","file, or use our example files."))),c.a.createElement(T.a,{className:"centered-content"},c.a.createElement("h3",{className:"example-text"},"See some examples:")),c.a.createElement(T.a,{className:"centered-content"},c.a.createElement(C.default.Group,{onChange:function(t){var a="/phylo-lib-web/examples/phyloXML/"+t.target.value+".xml";I(t.target.value),fetch(a).then((function(t){return t.blob()})).then((function(t){return O(t)}))},value:N},c.a.createElement(C.default.Button,{value:"amphi_frost"},"Amphibian Phylogeny"),c.a.createElement(C.default.Button,{value:"tol_156"},"The Tree of Life"),c.a.createElement(C.default.Button,{value:"apaf"},"Apaf-1 Gene Family Tree")),c.a.createElement(s.a,{content:c.a.createElement("p",null,"Example files obtained from"," ",c.a.createElement("a",{href:"http://phyloxml.org/",target:"_blank",rel:"noopener noreferrer"},"phyloxml.org"),".",""!==N?c.a.createElement("p",null,"See the phyloXML source for the current tree"," ",c.a.createElement("a",{href:"http://phyloxml.org/archaeopteryx-js/phyloxml_trees/"+N+".xml",target:"_blank",rel:"noopener noreferrer"},"here"),"."):null),title:"Credits",trigger:"click"},c.a.createElement("div",{className:"example-credits"},c.a.createElement(u.a,null)))),c.a.createElement(T.a,{className:"centered-content"},c.a.createElement("h3",{className:"example-text"},"Or upload your own PhyloXML file:")),c.a.createElement(T.a,{className:"horizontally-centered"},c.a.createElement(x.a,R,c.a.createElement(h.a,null,c.a.createElement(j.a,null),"Upload PhyloXML Files")),c.a.createElement(h.a,{onClick:function(){I(""),O(i)},className:"action-button",disabled:e.length<1},"Visualize")),f?c.a.createElement(T.a,{justify:"center"},c.a.createElement("div",{className:"ascii-phylo-container",ref:k},c.a.createElement("p",{className:"ascii-phylo"},b))):null))}var Ba=m.a.Content;function Fa(){var t=Object(n.useState)(!1),a=Object(G.a)(t,2),e=a[0],r=a[1],g=Object(n.useState)(""),o=Object(G.a)(g,2),i=o[0],l=o[1],A=Object(n.useState)([]),m=Object(G.a)(A,2),f=m[0],d=m[1],E=Object(n.useState)([]),v=Object(G.a)(E,2),b=v[0],y=v[1],w=Object(n.useState)([]),_=Object(G.a)(w,2),N=_[0],I=_[1],k=Object(n.useState)([]),S=Object(G.a)(k,2),O=S[0],R=S[1],D=Object(n.useRef)(null),P=function(t,a,e){try{var n=new FileReader;n.onload=function(){a.push(K(n.result)),O.length>2&&R(O.slice(-2)),2===a.length&&e&&M(a)},n.readAsText(t)}catch(c){console.log(c),console.log("File printing failed")}},X=function(){window.scrollTo({behavior:"smooth",top:D.current.offsetTop-15})},M=function(t){var a,e,n=function(t,a,e,n,c){var r=Y(t)+1|0,g=Y(a)+1|0,o=kt(t,a,e,n,c,r,g),i=yt.c(yt.c(o,r-1|0),g-1|0);return[St(t,a,o,e,n,c),i]}(t[0],t[1],1,-1,-1),c=Object(G.a)(n,1)[0],g=(a=c[0],e=c[1],Rt.contents="",Ot.contents=!1,Pt(a,e),Rt.contents);l(g.trim()),r(!0),null!==D.current?X():setTimeout((function(){X()}),1e3)},H={accept:".FASTA, .fasta, .txt",action:"https://www.mocky.io/v2/5cc8019d300000980a055e76",fileList:N,onChange:function(t){I(t.fileList)},headers:{authorization:"authorization-text"},multiple:!0,onRemove:function(t){O.splice(N.indexOf(t)),R(O)},transformFile:function(t){P(t,O,!1),N.push(t),I(N.slice(-2))}};return c.a.createElement("div",{className:"wrapper"},c.a.createElement(Ba,{justify:"center"},c.a.createElement(T.a,{className:"page",justify:"center",gutter:[16,16]},c.a.createElement("div",null,c.a.createElement("h1",null,"Visualize Pairwise DNA Alignments"),c.a.createElement("h2",null,"Visualize an"," ",c.a.createElement(p,{content:c.a.createElement("p",null,"A DNA sequence alignment arranges DNA to identify regions of similarity, helping reveal evolutionary relations between the sequences.",c.a.createElement("br",null),"A pairwise alignment is an alignment of 2 sequences, and PhyloML utilizes the Needleman-Wunsch algorithm to conduct this alignment."),vocab:"alignment",link:"https://www.sciencedirect.com/science/article/abs/pii/0022283670900574?via%3Dihub",linkText:"the original paper on the Needleman-Wunsch algorithm"})," ","of two DNA sequences. Begin by uploading two"," ",c.a.createElement(p,{content:c.a.createElement("p",null,"The FASTA format is a way of representing nucleotide sequences. It was popularized by the FASTA software developed by David J. Lipman and William R. Pearson."),vocab:".FASTA",link:"https://www.ncbi.nlm.nih.gov/pmc/articles/PMC280013/",linkText:"the original paper on the FASTA program"})," ","files, or use our example files."))),c.a.createElement(T.a,{className:"centered-content"},c.a.createElement("h3",{className:"example-text"},"See some examples:")),c.a.createElement(T.a,{className:"centered-content"},c.a.createElement(C.default.Group,{onChange:function(t){d([]);var a=t.target.value.split(",");y(a),a.map((function(t){fetch("/phylo-lib-web/examples/FASTA/"+t+".fasta").then((function(t){return t.blob()})).then((function(t){return P(t,f,!0)}))}))},value:b.join(",")},c.a.createElement(C.default.Button,{value:"h1n1,h3n2"},"H1N1 vs H3N2"),c.a.createElement(C.default.Button,{value:"h5n1,h7n7"},"H5N1 vs H7N7"),c.a.createElement(C.default.Button,{value:"mers,h7n9"},"MERS vs H7N9")),c.a.createElement(s.a,{content:c.a.createElement("div",null,c.a.createElement("p",null,"The pairwise alignment shown is a globally optimum alignment obtained via the Needleman-Wunch algorithm."),c.a.createElement("p",null,"Base pair matches, mismatches and gaps are visualized. A star indicates a pair match, a bar indicates a mismatch, and a blank space indicates a gap."),c.a.createElement("p",null,"Example files obtained from"," ",c.a.createElement("a",{href:"https://www.ncbi.nlm.nih.gov/nuccore/",target:"_blank",rel:"noopener noreferrer"},"NCBI Nucleotide"),".",2===b.length?c.a.createElement("p",null,"See the sources for the current alignment:"," ",c.a.createElement("a",{href:"/phylo-lib-web/examples/FASTA/"+b[0]+".fasta",target:"_blank",rel:"noopener noreferrer"},b[0]+".fasta")," ","and"," ",c.a.createElement("a",{href:"/phylo-lib-web/examples/FASTA/"+b[1]+".fasta",target:"_blank",rel:"noopener noreferrer"},b[1]+".fasta"),"."):null)),title:"Information and Credits",trigger:"hover"},c.a.createElement("div",{className:"example-credits"},c.a.createElement(u.a,null)))),c.a.createElement(T.a,{className:"centered-content"},c.a.createElement("h3",{className:"upload-text"},"Or upload 2 FASTA files:")),c.a.createElement(T.a,{className:"horizontally-centered"},c.a.createElement(x.a,H,c.a.createElement(h.a,null,c.a.createElement(j.a,null),"Upload .FASTA Files")),c.a.createElement(h.a,{onClick:function(){M(O),O.length>=2&&(d([]),y([]))},className:"action-button",disabled:O.length<2},"Display Alignment")),e?c.a.createElement(T.a,{justify:"center"},c.a.createElement("div",{className:"ascii-phylo-container",ref:D},c.a.createElement("p",{className:"ascii-phylo"},i))):null))}var La=function(t){var a,e=t.useBack,n=c.a.createElement("div",null,c.a.createElement("p",null,"A"," ",c.a.createElement("a",{href:"https://en.wikipedia.org/wiki/Phylogenetic_tree"},"phylogenetic tree")," ","is a tree that shows the evolutionary relationships between species. The structure of these trees can be stored in"," ",c.a.createElement("a",{href:"http://www.phyloxml.org/"},"XML")," files. To see visualizations of such trees, you can click Visualize Phylogenetic Trees."),c.a.createElement("p",null," ","We can also try to predict the structure of phylogenetic trees by looking at the similarities in the DNAs of different species. To know more, you can click Generate Phylogenetic Trees."));return c.a.createElement("header",{className:"site-header"},c.a.createElement(i.b,{className:"site-title",to:"/"},e?c.a.createElement(A.a,{style:{alignSelf:"center",paddingRight:"10px"}}):null,"PhyloML"),c.a.createElement(s.a,(a={className:"clickable",content:n},Object(o.a)(a,"className","tooltip"),Object(o.a)(a,"trigger","click"),a),c.a.createElement("a",{className:"tooltip"},c.a.createElement(u.a,{style:{paddingRight:"0.5em"}}),"Confused by all the terms?")))},Ua=function(t){var a=t.useBack,e=t.body;return c.a.createElement("div",null,c.a.createElement(La,{useBack:a}),e)};function za(){return c.a.createElement(i.a,null,c.a.createElement("div",null,c.a.createElement(l.c,null,c.a.createElement(l.a,{path:"/generate",component:function(){return c.a.createElement(Ua,{useBack:!0,body:c.a.createElement(ra,null)})}}),c.a.createElement(l.a,{path:"/visualize",component:function(){return c.a.createElement(Ua,{useBack:!0,body:c.a.createElement(Ha,null)})}}),c.a.createElement(l.a,{path:"/pairwise",component:function(){return c.a.createElement(Ua,{useBack:!0,body:c.a.createElement(Fa,null)})}}),c.a.createElement(l.a,{path:"/",component:function(){return c.a.createElement(Ua,{useBack:!1,body:c.a.createElement(_,null)})}}))))}Boolean("localhost"===window.location.hostname||"[::1]"===window.location.hostname||window.location.hostname.match(/^127(?:\.(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)){3}$/));g.a.render(c.a.createElement(c.a.StrictMode,null,c.a.createElement(za,null)),document.getElementById("root")),"serviceWorker"in navigator&&navigator.serviceWorker.ready.then((function(t){t.unregister()})).catch((function(t){console.error(t.message)}))},64:function(t,a,e){}},[[141,1,2]]]);
//# sourceMappingURL=main.7d5ef282.chunk.js.map